src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/PackagePropertyPage/PackageIconValueProvider.cs

```diff
@@ -0,0 +1,101 @@
+        private const string TrueValue = "True";
+
+        private readonly IProjectItemProvider _sourceItemsProvider;
+        private readonly UnconfiguredProject _unconfiguredProject;
+
+        [ImportingConstructor]
+        public PackageIconValueProvider(
+            [Import(ExportContractNames.ProjectItemProviders.SourceFiles)] IProjectItemProvider sourceItemsProvider,
+            UnconfiguredProject unconfiguredProject)
+        {
+            _sourceItemsProvider = sourceItemsProvider;
+            _unconfiguredProject = unconfiguredProject;
+        }
+
+        // https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#packageicon
+        private static string CreatePackageIcon(string filePath, string? packagePath)
+        {
+            string filename = Path.GetFileName(filePath);
+            // Make a slash-only value into empty string, so it won't get prepended onto the path.
+            packagePath = ("\\".Equals(packagePath, StringComparison.Ordinal) || "/".Equals(packagePath, StringComparison.Ordinal))
+                ? string.Empty : (packagePath ?? string.Empty);
+            // The assumption is that PackagePath does not contain a path to a file; only a directory path.
+            return Path.Combine(packagePath, filename);
+        }
+
+        private async Task<IProjectItem?> GetExistingNoneItemAsync(string existingPackageIcon)
+        {
+            foreach (IProjectItem noneItem in await _sourceItemsProvider.GetItemsAsync(NoneItem))
+            {
+                string pack = await noneItem.Metadata.GetEvaluatedPropertyValueAsync(PackPropertyName);
+                string itemIconFilename = Path.GetFileName(noneItem.EvaluatedInclude);
+                string existingIconFilename = Path.GetFileName(existingPackageIcon);
+                // Instead of doing pure equality between itemPackageIcon and existingPackageIcon, a user may update the PackagePath
+                // of the item and forget to update the PackageIcon to reflect those changes, or vice versa. Instead, if the filename
+                // of this packed None item and the filename of the PackageIcon match, consider those to be related to one another.
+                if (TrueValue.Equals(pack, StringComparison.OrdinalIgnoreCase) && itemIconFilename.Equals(existingIconFilename, StringComparison.Ordinal))
+                {
+                    return noneItem;
+                }
+            }
+
+            return null;
+        }
+
+        public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            string relativePath = PathHelper.MakeRelative(_unconfiguredProject, unevaluatedPropertyValue);
+            string existingPackageIcon = await defaultProperties.GetEvaluatedPropertyValueAsync(PackageIconPropertyName);
```

Comment:
Is it possible for `existingPackageIcon` to be the empty string? What happens if it is?

Reply:
It gets passed into `GetExistingNoneItemAsync`. That method uses it via `Path.GetFilename`. Just checked in .NET Fiddle, and `System.IO.Path.GetFileName(string.Empty) == string.Empty` (they are equal). So, if `existingPackageIcon` is an empty string, the logic will end up using it as an empty string (which is an acceptable path). The first time you add a package icon, it will be an empty string. Thus, `existingItem` (in the next line after this) will be `null`.